### PR TITLE
Add error calculation to interpolated points in absorption corrections

### LINKED ISSUE #29058: Add error calculation to interpolated points when running MonteCarloAbsorption
An error calculation was recently added into MonteCarloAbsorption in PR #28979. That change only added the error calculation onto bins in the output correction workspace that had been explicitly simulated ie not interpolated. This change is to add in an error calculation for any points in the output that have been obtained via interpolation. There are two different types of interpolation that can be used in the calculation:

a) wavelength interpolation. This is only used if the ResimulateTracksForDifferentWavelengths parameter is set to True which is not the default. The gsl routines for interpolation don't appear to return interpolation errors although seems like we only use gsl for spline interpolation. May require us to implement our own spline interpolation.

b) using a a sparse instrument. The spatial interpolation used in the sparse calculation means the errors need to be propagated through the interpolation calculation which needs further work. The inverse distance weighted approach to doing a 2D interpolation doesn't have a well defined interpolation error. Will look at changing to bilinear interpolation which has a known error calculation eg slides 9-10 of: https://www.win.tue.nl/casa/meetings/seminar/previous/_abstract030122_files/4.pdf

In both cases, once we have an interpolation error calculation (difference between interpolation function and real function) we would need to combine this with any random errors in the simulated data points that we're interpolating between

### PR DESCRIPTION
**Pull request by DannyHindson**
**Description of work.**

A previous change (#28979) added an error calculation into the MonteCarloAbsorption algorithm. This PR extends that work to also generate an error for any interpolated points. The Monte Carlo absorption calculation is able to perform the following types of interpolation:

1) wavelength interpolation within a single spectrum using either linear or cubic spline interpolation. This is required in cases where the simulation only generates absorption correction factors at a subset of the wavelength points in the spectrum (controlled via the NumberOfWavelengthPoints parameter in the MonteCarloAbsorption algorithm)

2) spatial interpolation of entire histograms between spectra. It previously used an inverse weighted distance method but it has been changed to use bilinear interpolation as part of this PR because the interpolation error is easier to calculate. This feature is referred to as "Sparse workspace" in the MonteCarloAbsorption algorithm and it is used on instruments with a v large number of detectors where a subset are simulated and the results for the rest of the detectors are obtained by interpolation. The spatial interpolation often also requires some interpolation of type 1) to be done as a final step because the range\bins of the histograms being interpolated may not line up.

For each type of interpolation the error calculation combines two source of error which are treated as independent (added in quadrature): 
- the errors in the simulated points are propagated through the relevant interpolation formula 
- an estimate of the interpolation error for linear and for spatial interpolation is calculated ie how well does the interpolation match the actual function. An interpolation error hasn't been calculated for the spline interpolation because it is complex to calculate (requires a 4th derivative) and it is expected to be small anyway.

The linear and spline interpolation was previously performed using a mix of native and GSL library functions. This has been changed to use Mantid implementations only because the GSL library didn't do the required error calculations.

*Bilinear interpolation*

The change to bilinear interpolation generates slightly different results to the inverse distance weighted approach (ie "y" changes as well as making it easier to calculate "e"). Both approaches use the 4 detectors in the sparse workspace that form a rectangle around the point where interpolation is requested. The biggest differences occur for points that lie halfway between between two detectors on the edge of the "rectangle". For bilinear interpolation the interpolation only includes contributions from the two detectors lying on that edge whereas the inverse distance weighted approach includes contributions from all 4 detectors still. In the ILLDirectGeometryReduction system test, most of the differences are 1-2% with a small number as large as ~10%. The approach of only using 2 of the 4 points seems more sensible so I have proceeded with this change.

*Future work*

There is an algorithm called SplineInterpolation that also does spline interpolation using the CubicSpline function that is available in the Mantid fitting functionality. This algorithm hasn't been updated as part of this PR because it's not used in the absorption correction functionality but it could be done in future. For some reason the Interpolate.cpp and CubicSpline.cpp units had separate\similar code that called GSL interpolation. This could be centralised to use the new native spline interpolation code written as part of this PR.

**To test:**

Run the following MonteCarloAbsorption calculation. It uses an instrument that has 25 detectors arranged in a square grid and it runs the Monte Carlo simulation on a sparse set of 3 x 3 detectors with data for the remaining 16 detectors being interpolated eg spectra numbers 1, 11, 3, 13 form a square of detectors where the simulation did a full calculation and data for spectrum 7 in the middle will be interpolated. Also within each spectrum only a subset of the wavelength points are simulated with the rest being interpolated.
Check the y values on the spatially interpolated spectra (eg workspace index 6, spectrum 7) look similar to the nearby spectra.
Observe the errors are now calculated for all the bins in all the spectra. If you plot a spectrum with error bars it should be possible to see the errors are slightly larger for the interpolated points.
Change the Interpolation parameter to 'Linear' and check the script works and that errors are still calculated for all bins on all spectra.

```
# import mantid algorithms, numpy and matplotlib
from mantid.simpleapi import *
import matplotlib.pyplot as plt
import numpy as np

data = CreateSampleWorkspace(WorkspaceType='Histogram', NumBanks=1, BankPixelWidth=5, XMin=1000)
data = ConvertUnits(data, Target='Wavelength')
SetSample(data, Geometry={'Shape': 'Cylinder', 'Height': 5.0, 'Radius': 1.0,
                  'Center': [0.0,0.0,0.0]},
                Material={'ChemicalFormula': 'V'})

abscor = MonteCarloAbsorption(data, SparseInstrument=True,
                              NumberOfDetectorRows=3, NumberOfDetectorColumns=3,
                              Interpolation='CSpline', # test also with this set to Linear
                              ResimulateTracksForDifferentWavelengths=True,
                              NumberOfWavelengthPoints=5)
corrected = data/abscor
```

Fixes #29058 

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

File: Framework/Algorithms/test/SparseWorkspaceTest.h [hunk 1]
```diff
@@ -298,6 +303,150 @@ class SparseWorkspaceTest : public CxxTest::TestSuite {
...
+    auto &grid = sparseWS->grid();
+    double lat = grid.latitudeAt(0);
+    double lon = grid.longitudeAt(0);
+    auto h = sparseWS->bilinearInterpolateFromDetectorGrid(lat, lon);
+    TS_ASSERT_EQUALS(h.size(), wavelengths)
+    for (size_t i = 0; i < h.size(); ++i) {
+      TS_ASSERT_EQUALS(h.y()[i], 0.0)
+      TS_ASSERT_EQUALS(h.e()[i], 0.0)
+    }
+    lat = (grid.latitudeAt(2) + grid.latitudeAt(1)) / 2.0;
```
Comment (Harrietbrown) [inline Framework/Algorithms/test/SparseWorkspaceTest.h:332]: Could this be split into two tests?

File: Framework/Algorithms/test/SparseWorkspaceTest.h [hunk 2]
```diff
@@ -298,6 +303,150 @@ class SparseWorkspaceTest : public CxxTest::TestSuite {
...
+    h = sparseWS->bilinearInterpolateFromDetectorGrid(lat, lon);
+    TS_ASSERT_EQUALS(h.size(), wavelengths)
+    for (size_t i = 0; i < h.size(); ++i) {
+      TS_ASSERT_DELTA(h.y()[i], val, 1e-7)
+      TS_ASSERT_DELTA(h.e()[i], err, 1e-7);
+    }
+  }
+
+  // test the interpolation error
+  void test_bilinearInterpolateFromDetectorGrid_interp_errors() {
```
Comment (Harrietbrown) [inline Framework/Algorithms/test/SparseWorkspaceTest.h:367]: This test is also very long. could it be split up?

File: Framework/HistogramData/src/Interpolate.cpp [hunk 3]
```diff
@@ -94,103 +103,249 @@ void sanityCheck(const Histogram &input, const Histogram &output,
...
+  auto &eold = input.dataE();
+  std::vector<double> u_ypp_ypp(xs.size());
+  std::vector<double> u_ypp_y(xs.size());
+
+  for (size_t i = 0; i < xs.size(); i++) {
+    for (size_t k = 0; k < xs.size(); k++) {
+      double dyppidyk = 0;
+      if ((i != 0) && (i != xs.size() - 1)) {
+        if (k > 1) {
+          dyppidyk += h[i - 1][k - 2] / (xs[k] - xs[k - 1]);
```
Comment (Harrietbrown) [inline Framework/HistogramData/src/Interpolate.cpp:161]: These variable names could do with being more clear what they are.

File: docs/source/algorithms/MonteCarloAbsorption-v1.rst [hunk 4]
```diff
@@ -112,31 +118,30 @@ When the sparse instrument option is enabled, a sparse instrument corresponding
...
-
-.. math::
-
-   \Delta_i = 2 \arcsin \sqrt{\sin^2 \left(\frac{\phi - \phi_i}{2} \right) + \cos \phi \cos \phi_i \sin^2 \left( \frac{\lambda - \lambda_i}{2} \right)}
+The sample to detector distance does not matter for absorption, so it suffices to consider directions only. The detector grid of the sparse instrument consists of detectors at constant latitude and longitude intervals. For a detector :math:`D` of the full input instrument at latitude :math:`\phi` and longitude :math:`\lambda`, we pick the four detectors :math:`D_ij` (:math:`i = 1, 2` :math:`j = 1, 2`) at the corners of the grid cell which includes (:math:`\phi`, :math:`\lambda`).
```
Comment (Harrietbrown) [inline docs/source/algorithms/MonteCarloAbsorption-v1.rst:125]:
```suggestion
The sample to detector distance does not matter for absorption, so it suffices to consider directions only. The detector grid of the sparse instrument consists of detectors at constant latitude and longitude intervals. For a detector :math:`D` of the full input instrument at latitude :math:`\phi` and longitude :math:`\lambda`, we pick the four detectors :math:`D_{ij}` (:math:`i = 1, 2` :math:`j = 1, 2`) at the corners of the grid cell which includes (:math:`\phi`, :math:`\lambda`).
```

**Review by Harrietbrown — CHANGES_REQUESTED** [review]
Very well writen, just a couple of notes and we can get this in by tomorrow. Good Job

File: Framework/Algorithms/test/SparseWorkspaceTest.h (hunk 1 above)
Comment (DannyHindson) [reply Framework/Algorithms/test/SparseWorkspaceTest.h:332]: Yes will do. I had based this rather long test on an existing test called test_interpolateFromDetectorGrid which used the original interpolation method. I will split up both and pull out the workspace creation into a separate function

File: Framework/HistogramData/src/Interpolate.cpp (hunk 3 above)
Comment (DannyHindson) [reply Framework/HistogramData/src/Interpolate.cpp:161]: Yes I'll see what I can do. I'll probably add a few comments to explain my naming convention eg dyppidyk is meant to be dy''_i/dy_k but was bit hindered by not being able to include certain characters in the variable names. So have used p to mean "prime"

**Review by Harrietbrown — APPROVED** [review]
Changes loog good, the notes on the meaning of the variable names are very helpful. Good Job

**Pull request merged**